src/plugins/intel_gpu/src/graph/impls/ocl/multi_stage_primitive.hpp

```diff
@@ -0,0 +1,226 @@
+    std::vector<kernel::ptr> _kernels;
+
+    // a pair of batch program hash and kernel entry hash of each ocl impl.
+    std::pair<std::string, std::string> kernel_dump_info;
+
+    multi_stage_primitive() : _kernels_data({}), _kernels({}) {}
+
+    multi_stage_primitive(const multi_stage_primitive<PType>& other)
+        : typed_primitive_impl<PType>()
+        , _kernels_data(other._kernels_data)
+        , _kernels({}) {
+        _kernels.reserve(other._kernels.size());
+        for (size_t k = 0; k < other._kernels.size(); ++k) {
+            _kernels.emplace_back(other._kernels[k]->clone());
+        }
+        this->can_reuse_memory = false;
+        this->_kernel_name = other._kernel_name;
+        this->_is_dynamic = other._is_dynamic;
+    }
+
+    multi_stage_primitive(const std::vector<kernel_selector::kernel_data>& kd)
+        : typed_primitive_impl<PType>()
+        , _kernels_data(kd) {
+        this->can_reuse_memory = false;
+        this->_kernel_name = kd[0].kernelName;
+    }
+
+    bool is_cpu() const final { return false; }
+
+    // Cache blob format:
+    //     [ kernel_selector::kernel_data ]
+    //     [ kernel_ids ]
+    void save(BinaryOutputBuffer& ob) const override {
+        primitive_impl::save(ob);
+        ob << _kernels_data.size();
+        for (auto& kd : _kernels_data) {
+            ob << make_data(&kd.internalBufferDataType, sizeof(kernel_selector::Datatype));
+            ob << kd.internalBufferSizes;
+            ob << kd.kernels;
+            ob << kd.kernelName;
+        }
+    }
+
+    void load(BinaryInputBuffer& ib) override {
+        primitive_impl::load(ib);
+        size_t kernels_size;
+        ib >> kernels_size;
+        _kernels_data.resize(kernels_size);
+        for (size_t i = 0; i < kernels_size; i++) {
+            kernel_selector::kernel_data kd;
+            ib >> make_data(&kd.internalBufferDataType, sizeof(kernel_selector::Datatype));
+            ib >> kd.internalBufferSizes;
+            ib >> kd.kernels;
+            ib >> kd.kernelName;
+            _kernels_data[i] = kd;
+        }
+    }
+
+protected:
+    virtual kernel_arguments_data get_arguments(const typed_primitive_inst<PType>& instance, size_t stage) const = 0;
+
+    event::ptr aggregate_events(const std::vector<event::ptr>& events, stream& stream, bool group = false, bool is_output = false) const {
+        if (events.size() == 1 && !is_output)
+            return events[0];
+
+        if (group && !is_output)
+            return stream.group_events(events);
+
+        return events.empty() ? stream.create_user_event(true)
+                              : stream.enqueue_marker(events, is_output);
+    }
+
+    void init_kernels(const kernels_cache& kernels_cache, const kernel_impl_params& params) override {
+        _kernels.clear();
+        if (!_kernels_data.empty() && !_kernels_data[0].kernels.empty()) {
+            auto compiled_kernels = kernels_cache.get_kernels(params);
```

Comment:
Maybe it makes sense to check if the total number of `kernels.size()` of all `_kernels_data` matches with `compiled_kernels.size()` - just to make sure we haven't missed something

Reply:
Added